File: src/brain/app/services/recovery_model.py
"""Decline-code-segmented recoverable-revenue estimator.

IMPORTANT — ALL RATES ARE INDUSTRY-INFORMED ESTIMATES, NOT GUARANTEES.
Sources: card-network published soft-decline guidance, publicly available
dunning-benchmark studies (Chargebee, Recurly, Paddle 2022-2024).  Actual
recovery depends on merchant category, issuer, retry timing, and card mix.
Do NOT present these numbers to customers as contractual commitments.

Decline category vocabulary (shared contract with portal and other agents):
  insufficient_funds   ISO 51
  do_not_honor         ISO 05 / generic
  transient            ISO 91, 96, processing_error — issuer temporarily unavailable
  velocity             ISO 65, 75 — card-limit / frequency exceeded
  expired_card         ISO 54
  invalid_data         ISO 14 — wrong card number / CVV / zip
  fraud                ISO 41, 43, 07 — lost/stolen/pickup  → DO NOT RETRY
  authentication       3DS / authentication_required
  unknown              anything not mapped above
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TypedDict


# ---------------------------------------------------------------------------
# Recovery rates by segment
# ---------------------------------------------------------------------------

class _SegmentConfig(TypedDict):
    rate: float
    do_not_retry: bool
    description: str


# Keys are the canonical category names used throughout the system.
SEGMENT_CONFIG: dict[str, _SegmentConfig] = {
    "insufficient_funds": {
        "rate": 0.35,
        "do_not_retry": False,
        "description": "ISO 51 — insufficient funds; retry after a few days (~35% est.)",
    },
    "do_not_honor": {
        "rate": 0.22,
        "do_not_retry": False,
        "description": "ISO 05 / generic decline (~22% est.)",
    },
    "transient": {
        "rate": 0.45,
        "do_not_retry": False,
        "description": "ISO 91/96/processing_error — issuer temporarily unavailable (~45% est.)",
    },
    "velocity": {
        "rate": 0.30,
        "do_not_retry": False,
        "description": "ISO 65/75 — frequency or limit exceeded; retry later (~30% est.)",
    },
    "expired_card": {
        "rate": 0.15,
        "do_not_retry": False,
        "description": "ISO 54 — card expired; requires customer card update (~15% est.)",
    },
    "invalid_data": {
        "rate": 0.08,
        "do_not_retry": False,
        "description": "ISO 14 — bad card number/CVV/zip; requires customer correction (~8% est.)",
    },
    "fraud": {
        "rate": 0.00,
        "do_not_retry": True,
        "description": "ISO 41/43/07 — lost/stolen/pickup/fraud; DO NOT RETRY (0%)",
    },
    "authentication": {
        "rate": 0.10,
        "do_not_retry": False,
        "description": "3DS / authentication_required; retry with step-up auth (~10% est.)",
    },
    "unknown": {
        "rate": 0.15,
        "do_not_retry": False,
        "description": "Unclassified decline — conservative fallback estimate (~15% est.)",
    },
}

# ---------------------------------------------------------------------------
# Default decline mix — used when the caller supplies only a total amount.
# Based on published industry benchmarks (2022-2024 SaaS dunning studies).
# These proportions are approximate; replace with merchant-specific data
# when available.
# ---------------------------------------------------------------------------
DEFAULT_DECLINE_MIX: dict[str, float] = {
    "insufficient_funds": 0.35,
    "do_not_honor":       0.20,
    "transient":          0.15,
    "velocity":           0.08,
    "expired_card":       0.07,
    "invalid_data":       0.06,
    "fraud":              0.05,
    "authentication":     0.04,
}
# Sanity-check: must sum to 1.0 (within float tolerance).
assert abs(sum(DEFAULT_DECLINE_MIX.values()) - 1.0) < 1e-9, "DEFAULT_DECLINE_MIX must sum to 1.0"


# ---------------------------------------------------------------------------
# Public data types
# ---------------------------------------------------------------------------

@dataclass
class SegmentEstimate:
    category: str
    amount_cents: int
    recovery_rate: float
    recoverable_cents: int
    do_not_retry: bool
    description: str


@dataclass
class RecoveryEstimate:
    recoverable_amount_cents: int
    blended_recovery_rate: float
    """Weighted average rate across all segments (0–1)."""
    confidence_low_cents: int
    """Lower bound of ±20% confidence band."""
    confidence_high_cents: int
    """Upper bound of ±20% confidence band."""
    segments: list[SegmentEstimate] = field(default_factory=list)
    used_default_mix: bool = False
    """True when the caller supplied only a total and we applied DEFAULT_DECLINE_MIX."""
    disclaimer: str = (
        "Recovery rates are industry-informed estimates, NOT guarantees. "
        "Actual recovery depends on merchant category, issuer, retry timing, and card mix."
    )


# ---------------------------------------------------------------------------
# Core estimator
# ---------------------------------------------------------------------------

def estimate_recovery(
    breakdown: dict[str, int] | None,
    total_amount_cents: int,
) -> RecoveryEstimate:
    """Compute the segmented recovery estimate.

    Parameters
    ----------
    breakdown:
        Optional mapping of {category: amount_cents}.  Categories must be
        keys from ``SEGMENT_CONFIG``.  Unknown categories are treated as
        ``"unknown"``.
    total_amount_cents:
        The aggregate failed amount.  Used as a fallback when ``breakdown``
        is None or empty, and also to catch callers who supply a breakdown
        that doesn't sum to total (we don't error — we trust the breakdown).
    """
    if not breakdown:
        # Caller didn't segment — apply the documented default mix.
        breakdown = {
            cat: int(total_amount_cents * share)
            for cat, share in DEFAULT_DECLINE_MIX.items()
        }
        used_default_mix = True
    else:
        used_default_mix = False

    segments: list[SegmentEstimate] = []
    total_recoverable = 0

    for cat, amount in breakdown.items():
        canonical = cat if cat in SEGMENT_CONFIG else "unknown"
        cfg = SEGMENT_CONFIG[canonical]
        recoverable = int(amount * cfg["rate"])
        total_recoverable += recoverable
        segments.append(
            SegmentEstimate(
                category=canonical,
                amount_cents=amount,
                recovery_rate=cfg["rate"],
                recoverable_cents=recoverable,
                do_not_retry=cfg["do_not_retry"],
                description=cfg["description"],
            )
        )

    total_input = sum(s.amount_cents for s in segments) or 1  # guard div-by-zero
    blended_rate = total_recoverable / total_input

    # ±20% confidence band (rough heuristic; tighten with real labeled data).
    low = int(total_recoverable * 0.80)
    high = int(total_recoverable * 1.20)

    return RecoveryEstimate(
        recoverable_amount_cents=total_recoverable,
        blended_recovery_rate=round(blended_rate, 4),
        confidence_low_cents=low,
        confidence_high_cents=high,
        segments=segments,
        used_default_mix=used_default_mix,
    )

File: src/portal/lib/recovery-model.ts
/**
 * Decline-category recovery model — the single source of truth the marketing
 * ROI estimator and the /api/audit route both use. Category vocabulary matches
 * the Syzm Brain (`recovery_model.py`). Rates are industry-informed ESTIMATES,
 * not guarantees.
 */

export type DeclineCategory =
  | "insufficient_funds"
  | "do_not_honor"
  | "transient"
  | "velocity"
  | "expired_card"
  | "invalid_data"
  | "fraud"
  | "authentication"
  | "unknown";

export const DECLINE_CATEGORIES: DeclineCategory[] = [
  "insufficient_funds",
  "do_not_honor",
  "transient",
  "velocity",
  "expired_card",
  "invalid_data",
  "fraud",
  "authentication",
  "unknown",
];

export const CATEGORY_LABEL: Record<DeclineCategory, string> = {
  insufficient_funds: "Insufficient funds",
  do_not_honor: "Do not honor",
  transient: "Issuer transient / unavailable",
  velocity: "Velocity / limit exceeded",
  expired_card: "Expired card",
  invalid_data: "Invalid data",
  fraud: "Fraud / lost / stolen",
  authentication: "Authentication required",
  unknown: "Uncategorized",
};

/** Estimated share of failed-payment dollars recoverable via well-timed retries. */
export const RECOVERY_RATE: Record<DeclineCategory, number> = {
  insufficient_funds: 0.35,
  do_not_honor: 0.22,
  transient: 0.45,
  velocity: 0.3,
  expired_card: 0.15,
  invalid_data: 0.08,
  fraud: 0.0,
  authentication: 0.1,
  unknown: 0.15,
};

/** Categories Syzm will never retry (scheme rules + fraud risk). */
export const DO_NOT_RETRY: DeclineCategory[] = ["fraud"];

/** Default distribution of failed dollars across categories (industry-informed). Sums to 1. */
export const DEFAULT_MIX: Record<DeclineCategory, number> = {
  insufficient_funds: 0.32,
  do_not_honor: 0.19,
  transient: 0.12,
  velocity: 0.06,
  expired_card: 0.12,
  invalid_data: 0.05,
  fraud: 0.06,
  authentication: 0.04,
  unknown: 0.04,
};

export interface RecoverySegment {
  category: DeclineCategory;
  label: string;
  failedCents: number;
  rate: number;
  recoverableCents: number;
}

export interface RecoveryEstimate {
  failedCents: number;
  recoverableCents: number;
  blendedRate: number;
  band: { lowCents: number; highCents: number };
  segments: RecoverySegment[];
}

/** Map a raw processor decline code/string to a Syzm category. Best-effort. */
export function mapDeclineCode(raw: string): DeclineCategory {
  const c = raw.trim().toLowerCase();
  if (!c) return "unknown";
  if (/(insufficient|nsf|51|\b13\b)/.test(c)) return "insufficient_funds";
  if (/(do[_ ]?not[_ ]?honor|generic|05|do_not_honor)/.test(c)) return "do_not_honor";
  if (/(try[_ ]?again|issuer[_ ]?unavailable|processing[_ ]?error|reenter|91|96|server)/.test(c)) return "transient";
  if (/(velocity|limit|exceed|65|75|61)/.test(c)) return "velocity";
  if (/(expire|54)/.test(c)) return "expired_card";
  if (/(invalid|incorrect|14|15|format)/.test(c)) return "invalid_data";
  if (/(fraud|lost|stolen|pickup|pick[_ ]?up|41|43|07|57)/.test(c)) return "fraud";
  if (/(auth|3ds|sca|secure)/.test(c)) return "authentication";
  return "unknown";
}

export function blendedRate(mix: Record<DeclineCategory, number> = DEFAULT_MIX): number {
  return DECLINE_CATEGORIES.reduce((sum, c) => sum + mix[c] * RECOVERY_RATE[c], 0);
}

/**
 * Estimate recoverable revenue from total failed cents, split across categories
 * via `mix` (pass a real per-category breakdown for precision).
 */
export function estimateRecoverable(
  failedCents: number,
  mix: Record<DeclineCategory, number> = DEFAULT_MIX,
  bandFraction = 0.25,
): RecoveryEstimate {
  const safe = Number.isFinite(failedCents) && failedCents > 0 ? Math.trunc(failedCents) : 0;

  const segments: RecoverySegment[] = DECLINE_CATEGORIES.map((category) => {
    const failed = Math.round(safe * (mix[category] ?? 0));
    const rate = RECOVERY_RATE[category];
    return {
      category,
      label: CATEGORY_LABEL[category],
      failedCents: failed,
      rate,
      recoverableCents: Math.round(failed * rate),
    };
  }).filter((s) => s.failedCents > 0);

  const recoverableCents = segments.reduce((s, seg) => s + seg.recoverableCents, 0);
  const blended = safe > 0 ? recoverableCents / safe : 0;

  return {
    failedCents: safe,
    recoverableCents,
    blendedRate: blended,
    band: {
      lowCents: Math.round(recoverableCents * (1 - bandFraction)),
      highCents: Math.round(recoverableCents * (1 + bandFraction)),
    },
    segments,
  };
}

/** Build a normalized category mix from raw per-category failed-dollar totals. */
export function mixFromTotals(totals: Partial<Record<DeclineCategory, number>>): Record<DeclineCategory, number> {
  const sum = DECLINE_CATEGORIES.reduce((s, c) => s + (totals[c] ?? 0), 0);
  const mix = {} as Record<DeclineCategory, number>;
  for (const c of DECLINE_CATEGORIES) mix[c] = sum > 0 ? (totals[c] ?? 0) / sum : DEFAULT_MIX[c];
  return mix;
}
